### PR TITLE
Fix predicate elimination

### LINKED ISSUE #1571: Predicate elimination bug
This logic seems wrong (blaming myself):

https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/lower_predicate.cpp#L343-L348
```
filters.emplace_back([this](Expr* expr) {
    if (expr->isA<ReductionOp>()) {
      auto input = expr->inputs()[0]->as<TensorView>();
      auto input_def = input->definition();
      // When input_def is null, input must be an input to the fusion,
      // so that must be allocated on global memory. Since we don't omit
      // predication for expressions involving global memory, this
      // should never occur.
      TORCH_INTERNAL_ASSERT(
          input_def != nullptr, "Inconsistent input found: ", input);

      if (non_predicated_exprs_.find(input_def) !=
              non_predicated_exprs_.end() &&
          !(input_def->isA<ReductionOp>() &&
            (expr->as<ReductionOp>()->getReductionOpType() ==
             input_def->as<ReductionOp>()->getReductionOpType()))) {
        return true;
      }
    }
    return false;
  });
```

Repro:

```
TEST_F(NVFuserTest, FusionPredicateElimination2_CUDA) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  std::vector<int64_t> shape({10, 11});

  auto tv0 = makeConcreteTensor(shape);
  fusion.addInput(tv0);

  auto tv1 = add(tv0, IrBuilder::create<Double>(1));
  auto tv2 = sum(tv1, {1});
  auto tv3 = add(tv2, IrBuilder::create<Double>(1));

  fusion.addOutput(tv3);

  tv1->split(1, 4);
  tv1->split(0, 4);
  tv2->split(1, 4);
  tv2->split(0, 4);

  fusion.printMath();
  fusion.printKernel();

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  auto t0 = at::randn(shape, options);

  FusionExecutor fe;
  fe.compileFusion(&fusion, {t0});
  auto cg_outputs = fe.runFusion({t0});

  auto ref = (t0 + 1).sum({1}) + 1;

  testValidate(&fusion, cg_outputs, {t0}, {ref}, __LINE__, __FILE__);
}
```

Validation fails:

```
Validation error in output 0 on line 17049 in file ../torch/csrc/jit/codegen/cuda/test/test_gpu.cpp.
  Detected abs error of: 1.81327
    absolute tolerance was set to 4.4778e-06
    and relative tolerance set to 4.4778e-08
Exception raised from testValidate at ../torch/csrc/jit/codegen/cuda/test/test_gpu_validator.h:436 (most recent call first):
```

tv2's predicate is eliminated, which is wrong:

```
  for(nvfuser_index_t i24 = 0; i24 < (ceilDiv(10, 4)); ++i24) {
    #pragma unroll
    for(nvfuser_index_t i25 = 0; i25 < 4; ++i25) {
      int64_t i106;
      i106 = (i24 * 4) + i25;
      #pragma unroll
      for(nvfuser_index_t i26 = 0; i26 < (ceilDiv(11, 4)); ++i26) {
        #pragma unroll
        for(nvfuser_index_t i27 = 0; i27 < 4; ++i27) {
          T2[i106]
            = T2[i106]
            + T1[(i106 * 11) + ((i26 * 4) + i27)];
        }
      }
    }
  }
```

### PR DESCRIPTION
**Pull request by naoyam**
Fixes #1571

Revisited the predicate elimination logic. As mentioned in #1571, the change in index linearlization required more strict conditions, although uncommon. Also supported elimination predicates of welford ops.

This PR mainly just fixes the current issues. I'll have a follow-up PR to clean it up.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_predicate.cpp [hunk 1]
```diff
@@ -256,6 +256,32 @@ class PredicateAnalyzer : public OptOutDispatch {
   bool needs_predicate_ = false;
 };
 
+// Returns the initialization value of tv or nullptr if not initialized.
+Val* getInitVal(TensorView* tv) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_predicate.cpp]: can this go in ir_utils, seems generally useful. Maybe phrase it more like `getReductionInitValOf`

File: torch/csrc/jit/codegen/cuda/predicate_compute.cpp [hunk 2]
```diff
@@ -436,6 +436,7 @@ void UnswitchPredicate::predicateOn(Expr* tv_expr) {
 
   const auto gpu_lower = GpuLower::current();
   if (gpu_lower->predicateElimination().canOmitPredicate(tv_expr)) {
+    addParallelizedDomainPredicates(tv_expr);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/predicate_compute.cpp:439]: This change is to add predicates for parallel domains (e.g., `threadIdx.x < N`), even when tensors do not need predicates.

File: torch/csrc/jit/codegen/cuda/lower_predicate.cpp [hunk 3]
```diff
@@ -324,17 +350,18 @@ bool PredicateElimination::needsPredicate(Expr* expr) const {
...
-  // pattern with rfactor, the input should be safe to use with no
-  // predication.
+  // for this reduction. However, it is still legal to omit its
+  // predicate when: 1) the predicate of the input is not omitted and
+  // 2) the input can be initialized to the init value of this
+  // reduction. When the input is the output of another reduciton, the
+  // input is initialized to the init value of the reduction, so the
+  // two reductions must use the same init value.
+  // See FusionPredicateElimination3 and FusionPredicateElimination4
+  // for concrete examples.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_predicate.cpp:336]: I'll refactor this function in a follow-up PR. I don't like to have this large function, although it's conceptually separated into a series of anonymous lambdas.

File: torch/csrc/jit/codegen/cuda/lower_predicate.cpp [hunk 4]
```diff
@@ -344,11 +371,198 @@ bool PredicateElimination::needsPredicate(Expr* expr) const {
...
+    return false;
+  });
+
+  // An index can exceed the logical extent of the indexed domain if
+  // it's split. It can cause a reduction op to reduce the same value
+  // multiple times. Even a pointwise op can be a problem if the
+  // consumer is an alias of the producer. This check excludes such
+  // expressions from predicate elimination.
+  //
+  // This is not an issue if the index includes a zero domain (as defined in
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_predicate.cpp:459]: I think this is also not an issue if all domains are within the exact map across producer and consumers.

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/lower_predicate.cpp (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_predicate.cpp:459]: This is a wrong statement, forget it.

**Pull request merged**